### ISSUE TITLE
database.php script during initial install fails

### ISSUE THREAD

**Issue by Farmbuyer** [filesender/filesender#1323]
Installing 2.35 on Ubuntu 22.04 using nginx, PHP 8.1, and PostgreSQL 14.  Step 9 has us run the database.php script; running as the unprivileged user gives endless errors, and running as root triggers SELinux.  Running as the www-data user seems to work, but then fails with

```
Checking class AggregateStatistic
Working on table AggregateStatistics
Look for table AggregateStatistics
Table found, check columns
Found 7 columns in existing table : epoch, epochtype, eventtype, eventcount, timesum, sizesum, encryptedsum
Found 7 columns in required table : epoch, epochtype, eventtype, eventcount, timesum, sizesum, encryptedsum
Check column format
        timesum default is not "0"
Column timesum has bad format, updating it
Error, Rolling database changes back....
 This should leave the database state as it was before you started the script
Encountered exception : database_access_failure, see logs for details (uid: 6328be58bd2b0) ...\n
```

The end of /var/log/postgresql/postgresql-14-main.log is
```
2022-09-19 19:07:06.273 UTC [186698] LOG:  database system is ready to accept connections
2022-09-19 19:09:12.770 UTC [186857] filesender@filesender ERROR:  could not determine data type of parameter $1
2022-09-19 19:09:12.770 UTC [186857] filesender@filesender STATEMENT:  ALTER TABLE AggregateStatistics ALTER COLUMN timesum SET DEFAULT $1
```

The contents of /opt/filesender/filesender/log/filesender_2022-09-19_20h.log may be a little verbose for pasting into an issue; if this is too much let me know I can delete this text in favor of attaching the logfile:
```
[2022-09-19 20:09:12] [upgrade:error] [8192] preg_split(): Passing null to parameter #3 ($limit) of type int is deprecated in /opt/filesender/filesender/classes/autoload.php at line 166
[2022-09-19 20:09:12] [upgrade:error] [8192] preg_split(): Passing null to parameter #3 ($limit) of type int is deprecated in /opt/filesender/filesender/classes/autoload.php at line 166
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException exception uid:6328be58bd2b0] database_access_failure
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException trace uid:6328be58bd2b0] #0 /opt/filesender/filesender/classes/utils/DatabasePgsql.class.php(546): DBIStatement->__call()
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException trace uid:6328be58bd2b0] #1 /opt/filesender/filesender/classes/utils/Database.class.php(149): DatabasePgsql::updateTableColumnFormat()
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException trace uid:6328be58bd2b0] #2 /opt/filesender/filesender/scripts/upgrade/database.php(459): Database::updateTableColumnFormat()
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException trace uid:6328be58bd2b0] #3 /opt/filesender/filesender/scripts/upgrade/database.php(178): updateTable()
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException trace uid:6328be58bd2b0] #4 /opt/filesender/filesender/scripts/upgrade/database.php(222): ensureTable()
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException trace uid:6328be58bd2b0] #5 /opt/filesender/filesender/scripts/upgrade/database.php(688): ensureAllTables()
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException trace uid:6328be58bd2b0] #6 {main}
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0] "SQLSTATE[42P18]: Indeterminate datatype: 7 ERROR:  could not determine data type of parameter $1"
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0] Array
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0] (
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]     [method] => execute
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]     [args] => Array
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]         (
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]             [0] => Array
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]                 (
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]                     [:default] => 0
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]                 )
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]         )
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]     [query] => ALTER TABLE AggregateStatistics ALTER COLUMN timesum SET DEFAULT :default
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0] )
[2022-09-19 20:09:12] [upgrade:error] [DBIUsageException details uid:6328be58bd2b0]
```

Absolutely no idea what to do with that information, I'm afraid.  It's been approximately forever since I've read or written any SQL myself.  Is this an error that's going to cause problems further down the line, or is it something we can ignore since -- as a fresh installation -- there was no previous database present to upgrade?

**Comment by madsi1m** [comment]
Hi, try php 7.3 or php 7.4 see if it still fails.

**Comment by monkeyiq** [comment]
Thank you for the report.

Does psql allow you to run the SQL directly?
```
ALTER TABLE AggregateStatistics ALTER COLUMN timesum SET DEFAULT 0
```

If you are creating an installation you could try running `database.php` on an empty database as the www-data user and see if it completes. 

I have here
```
rpm -q php postgresql
php-8.1.9-1.fc36.x86_64
postgresql-14.3-2.fc36.x86_64
```

**Comment by Farmbuyer** [comment]
> Does psql allow you to run the SQL directly?

Maybe?  I can paste exactly that line into a psql prompt with the 'filesender' database opened by the 'filesender' user, and it gives no output.  I'm a Unix-y kind of guy, so when it's completely silent I'll assume that means success.  :-)  I don't know how to query the appropriate data to test what the new state of things is, though.

> If you are creating an installation you could try running database.php on an empty database as the www-data user and see if it completes.

If it's a brand-new installation, wouldn't the database already be empty?  (Not a rhetorical question, I don't really do DBA type stuff.)

We already have to run the database.php script as www-data (I used `sudo -u www-data php .....`) otherwise it can't open the config file, the logs directory, etc; that's how I got the output in the original post.

Does the PostgreSQL server know about 'www-user'?  My extremely sketchy understanding is that the username/uid running the psql or similar client process doesn't really come into play unless "peer" authentication is used.  The installation instructions have us changing the auth method from peer to md5, so I would have thought that permissions _inside_ the database all look to the 'filesender' username, not the Unix username?

(The installation instructions' example block of `pg_hba.conf` seems to be missing some newlines, by the way.  That took me a little while to figure out, heh.)

**Comment by Nicodonald** [comment]
Same problem here with Ubuntu 22.04
PHP 8.1.2
Posgresql 14+238

I tried to alter the column default to 0 and it worked (or it was already 0)
```
filesender=> \d AggregateStatistics
                        Table "public.aggregatestatistics"
    Column    |            Type             | Collation | Nullable |   Default
--------------+-----------------------------+-----------+----------+--------------
 epoch        | timestamp without time zone |           | not null |
 epochtype    | integer                     |           | not null |
 eventtype    | integer                     |           | not null |
 eventcount   | bigint                      |           |          |
 timesum      | numeric                     |           | not null | '0'::numeric
 sizesum      | numeric                     |           | not null | '0'::numeric
 encryptedsum | numeric                     |           | not null | '0'::numeric
```
and I reran the script but same error
```
Look for table AggregateStatistics
Table found, check columns
Found 7 columns in existing table : epoch, epochtype, eventtype, eventcount, timesum, sizesum, encryptedsum
Found 7 columns in required table : epoch, epochtype, eventtype, eventcount, timesum, sizesum, encryptedsum
Check column format
        timesum default is not "0"
Column timesum has bad format, updating it
Error, Rolling database changes back....
 This should leave the database state as it was before you started the script
Encountered exception : database_access_failure, see logs for details (uid: 636c0e5faa937)
```

**Comment by monkeyiq** [comment]
Please let me know if https://github.com/filesender/filesender/pull/1347 fixes this for you. This will be merged into development soon (perhaps by the time you read this) so you could then just pull development to test.

**Comment by monkeyiq** [comment]
This should be part of release 2.39.

**Comment by WebSpider** [comment]
@Farmbuyer can you check if this is fixed for you in 2.39+ ?

**Comment by Farmbuyer** [comment]
@WebSpider sorry, GitHub apparently doesn't give me any notifications on "@" comments, so I'm only now seeing this.  I'll try to get a new download for testing sometime in the next week!
